presto-docs/src/main/sphinx/admin/cte-materialization.rst

```diff
@@ -0,0 +1,103 @@
+=================================
+CTE Materialization
+=================================
+
+CTEs (Common Table Expressions) are defined as subqueries that appear in a WITH clause provided by the user.
+Their repeated usage within a query can lead in unnecessary redundant computations, excessive data retrieval, and high resource consumption
+
+To address this, Presto supports CTE Materialization. This feature allows intermediate CTEs to be materialized and effectively stored
+and reused within the scope of the same query.
+
+This improves performance and is especially beneficial when the same CTE is utilized multiple times in a query.
+These CTEs will be stored in temporary tables that are bucketed based on the first projection column
+
+To use this feature the connector used by the query must support the creation of temporary tables. Currently, only the :doc:`/connector/hive` connector offers this capability.
+
+How to use CTE Materialization
+--------------
+
+The following configurations and session properties are provided to enable CTE materialization and fine-tune its settings.
+
+``cte-materialization-strategy``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Allowed values:** ``ALL``, ``NONE``, ``HEURISTIC`, ``HEURISTIC_COMPLEX_QUERIES_ONLY``
+    * **Default value:** ``NONE``
+
+    Specifies the strategy to use for materializing Common Table Expressions (CTEs) in queries.
+    ``NONE`` indicates that no CTEs will be materialized.
+    ``ALL`` indicates that all CTEs in the query will be materialized.
+    ``HEURISTIC` greedily materializes the earliest parent CTE which is repeated >= cte_heuristic_replication_threshold times
+    ``HEURISTIC_COMPLEX_QUERIES_ONLY` greedily materializes the earliest parent CTE which meets the heuristic criteria and has a join or aggregate
+    This can also be specified on a per-query basis using the ``cte_materialization_strategy`` session property.
+
+``cte-heuristic-replication-threshold``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Minimum value:** ``0``
+    * **Default value:** ``4``
+
+    Used with CTE Materialization Strategy is set to ``HEURISTIC` or ``HEURISTIC_COMPLEX_QUERIES_ONLY``. CTES are only materialized if they are used greater than or equal to this number
+    This can also be specified on a per-query basis using the ``cte_heuristic_replication_threshold`` session property.
+
+``query.cte-hash-partition-count``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Default value:** ``100``
+
+    The number of partitions to be used for materializing Common Table Expressions (CTEs) in queries.
+    This setting determines how many buckets or writers should be used when materializing the CTEs, potentially affecting the performance of queries involving CTE materialization.
+    A higher number of partitions might improve parallelism but also increases overhead in terms of memory and network communication.
```

Comment:
Its strange that a property called partitioning actually affects the buckets. But please emphasize in the explanation that these are buckets and not partitions.

Reply:
This actually comes from exchange materialization where we use the hash partition count to count the number of buckets

Reply:
Thanks for explaining! I thought this was strange, but the explanation makes sense.